src/model/layer/opencl_convolution.cc

```diff
@@ -0,0 +1,256 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+#include "opencl_convolution.h"
+
+#ifdef USE_OPENCL
+
+namespace singa {
+
+RegisterLayerClass(OpenclConvolution);
+
+
+/// \copydoc Layer::Forward(int flag, const Tensor&)
+const Tensor OpenclConvolution::Forward(int flag, const Tensor &input) {
+  CHECK(buf_.empty());
+  CHECK_EQ(input.device()->lang(), kOpencl);
+  CHECK_EQ(input.nDim(), 4u);
+  
+  if (flag & kTrain) buf_.push(input);
+  
+  auto batchsize = input.shape(0);
+  auto imagesize = input.Size() / batchsize;
+  auto data_type = input.data_type();
+  auto device = input.device();
+  
+  Shape shape{batchsize, num_filters_, conv_height_, conv_width_};
+  Tensor output(shape, device, data_type);
+  Tensor col_data(Shape{col_height_, col_width_}, device, data_type);
+  
+  for (size_t b = 0; b < batchsize; b++) {
```

Comment:
opencl process all images in one batch in parallel.
never mind. let's do the optimization later.